### PR TITLE
Extensible Starknet Runtime type boilerplate

### PR DESCRIPTION
**Pull request by MaksymilianDemitraszek**
Related #916 

## Introduced changes

- Introduces type boilerplate for the extension system, future PRs will migrate existing codebase to use this boilerplate.

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Review by war-in — COMMENTED** [review]
Would be nice to see it in action (some migrated code). Currently, I can't see what it simplifies

**Review by piotmag769 — COMMENTED** [review]
I really like this, but I think it would be better to do a stack of PRs implementing this trait for both forge and cast scripts so we are not left with a dead code that will be forgotten

**Comment by MaksymilianDemitraszek** [comment]
@piotmag769 yea I have another PR wip wich migrates stuff partially

File: crates/forge-runner/src/runtime.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+impl<'a> HintProcessorLogic for StarknetRuntime<'a> {
+    fn execute_hint(
+        &mut self,
+        vm: &mut VirtualMachine,
+        exec_scopes: &mut ExecutionScopes,
+        hint_data: &Box<dyn Any>,
+        constants: &HashMap<String, Felt252>,
+    ) -> Result<(), HintError> {
```
Comment (Arcticae) [inline crates/forge-runner/src/runtime.rs:51]:
```suggestion
    ) -> HintExecutionResult {
```
is that correct?
would make for clearer code

File: crates/forge-runner/src/runtime.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+#[allow(dead_code)]
+enum HintHandlingResult {
+    Forward,
+    Result(HintExecutionResult),
+}
+
+#[allow(dead_code)]
+enum CheatcodeHadlingResult {
```
Comment (Arcticae) [inline crates/forge-runner/src/runtime.rs:122]:
```suggestion
enum CheatcodeHandlingResult {
```

File: crates/forge-runner/src/runtime.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+trait ExtensionLogic {
+    fn override_system_call(_syscall: SyscallSelector) -> HintHandlingResult {
+        HintHandlingResult::Forward
+    }
+
+    fn handle_cheatcode(_selector: &str, _inputs: Vec<Felt252>) -> CheatcodeHadlingResult {
+        CheatcodeHadlingResult::Forward
+    }
```
Comment (Arcticae) [inline crates/forge-runner/src/runtime.rs:134]: Hard to say what will happen after this PR but you might be in a similar situation that i was, where separate Extensions would require cross-extension state access. Did you take that into account?

File: crates/forge-runner/src/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        reference_ids: &HashMap<String, usize>,
+        references: &[HintReference],
+    ) -> Result<Box<dyn Any>, VirtualMachineError> {
+        self.hint_handler
+            .compile_hint(hint_code, ap_tracking_data, reference_ids, references)
+    }
+}
+
+pub struct ExtendedRuntime<Handler, Runtime: HintProcessor> {
+    pub extension_handler: Handler,
```
Comment (piotmag769) [inline crates/forge-runner/src/runtime.rs:69]: Actually what is the purpose of this one? It is not enforce to implement any trait for it, I think it needs a constraint

**Pull request closed without merging**